### PR TITLE
feat: more flexible JSON parsing

### LINKED ISSUE #8326: feat: better JSON parsing
### Component

Forge

### Describe the feature you would like

Current UX for JSON-parsing of objects is not great. Users are required to place JSON fields in alphabetical order, and `parseJson` basically guesses types, making it impossible to parse e.g. fixed-length arrays or `bytes` of length 20 or 32. 

To fix this we need to make JSON parser aware of struct field types and names. This issue proposes the following approach.

1. Add new cheatcodes
    ```solidity
    function parseJsonStruct(string calldata json, string calldata schema) external pure returns (bytes memory abiEncodedData);
    
    function parseJsonStruct(string calldata json, string calldata key, string calldata schema) external pure returns (bytes memory abiEncodedData);
    ```
    
    Those cheatcodes will be identical to `parseJson`, but will be guided by types from `schema`. Format for `schema` would be the json representation of `Vec<StructField>` where `StructField` is:
    ```rust
    struct StructField {
         /// Name of the field which will be used when parsing JSON.
         name: String,
         /// Type of the field
         ty: StructFieldType,
    }
    
    /// Solidity type representation which can be (de)serialized from Json and converted into DynSolType
    enum StructFieldType {
        Struct(Vec<StructField>),
        Array(Box<StructFieldType>),
        FixedArray(Box<StructFieldType>, usize),
        /// Inner value must be decodable into DynSolType.
        Primitive(String),
    }
    ```
    
    Another way to represent this in JSON could be by using `alloy_dyn_abi::Resolver` used for EIP-712 + type name.
2. Add helpers for generating `schema` values. For example, we could add `forge bind-json` command which would accept a path to .sol file, and produce either a schema for all structs, or a comlete parsing library looking like:
```solidity
library Helpers {
    string constant SCHEMA_MyStruct = '[{"name": "field1", "ty": "uint256"}, ...]';
    string constant SCHEMA_AnotherStruct = ...;

    function parseMyStruct(string memory json) internal pure returns(MyStruct memory) {
        return abi.decode(vm.parseJsonStruct(json, SCHEMA_MyStruct), (MyStruct));
    }
    ...
}
``` 

The same approach can be used for serialization of structs as well:
```solidity
function serializeStruct(string calldata schema, bytes memory abiEncodedData)
        external
        returns (string memory json);

function serializeStruct(string calldata objectKey, string calldata valueKey, string calldata schema, bytes memory abiEncodedData)
        external
        returns (string memory json);
```

Such approach reduces compilation overhead by keeping most of the logic in cheatcode implementation, and only requiring contracts to only contain relatively small schema definitions.

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

closes #8326 

## Solution

Adds `parseJsonType` set of cheatcodes which accepts stringified Solidity type as guidance for the parser, allowing to express complex types without the need to resort to parsing of them in multiple `parseJson*` calls.

Example usage of such cheatcode:
```solidity
// for simple types
uint256 value = abi.decode(vm.parseJsonType("100", "uint256"), (uint256)));
// for more complex stuff
address[][3] memory values = abi.decode(vm.parseJsonType('[[],["0xd8da6bf26964af9d7eed9e03e53415d37aa96045"], []]', "address[][3]"), (address[][3]));

struct Mail {
    address from;
    address to;
    string contents;
}

// struct types are encoded as EIP-712 `encodeType`
// fields order doesn't matter
Mail memory mail = abi.decode(
    vm.parseJsonType(
        '{"from": "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045", "contents": "something", "to": "0xEe5F5c53CE2159fC6DD4b0571E86a4A390D04846"}',
        "Mail(address from, address to, string contents)"
    ),
    (Mail)
);


```

For simpler creation of type strings for structs added `forge eip712 <path>` command which creates EIP-712 `encodeType` string for all structs in the given file.

Example output:
```
forge eip712 lib/forge-std/src/Vm.sol   
[⠊] Compiling...
[⠢] Compiling 1 files with Solc 0.8.26
[⠆] Solc 0.8.26 finished in 156.26ms
No files changed, compilation skipped
Log(bytes32[] topics,bytes data,address emitter)

Rpc(string key,string url)

EthGetLogs(address emitter,bytes32[] topics,bytes data,bytes32 blockHash,uint64 blockNumber,bytes32 transactionHash,uint64 transactionIndex,uint256 logIndex,bool removed)

DirEntry(string errorMessage,string path,uint64 depth,bool isDir,bool isSymlink)

FsMetadata(bool isDir,bool isSymlink,uint256 length,bool readOnly,uint256 modified,uint256 accessed,uint256 created)

Wallet(address addr,uint256 publicKeyX,uint256 publicKeyY,uint256 privateKey)

FfiResult(int32 exitCode,bytes stdout,bytes stderr)

ChainInfo(uint256 forkId,uint256 chainId)

AccountAccess(ChainInfo chainInfo,uint8 kind,address account,address accessor,bool initialized,uint256 oldBalance,uint256 newBalance,bytes deployedCode,uint256 value,bytes data,bool reverted,StorageAccess[] storageAccesses,uint64 depth)ChainInfo(uint256 forkId,uint256 chainId)StorageAccess(address account,bytes32 slot,bool isWrite,bytes32 previousValue,bytes32 newValue,bool reverted)

StorageAccess(address account,bytes32 slot,bool isWrite,bytes32 previousValue,bytes32 newValue,bool reverted)

Gas(uint64 gasLimit,uint64 gasTotalUsed,uint64 gasMemoryUsed,int64 gasRefunded,uint64 gasRemaining)
```

This might also be useful for https://github.com/foundry-rs/foundry/issues/4818 later

Will be working on serialization next

### CODE REVIEW

**Comment by klkvr** [comment]
Advantage of using EIP-712 is that we can reuse some of existing alloy structures for it and it's also pretty much human readable and compact.

However, it has some limitations 
1. We can't represent enums with it. Currently enums are treated as `uint8` but we could allow (de)serialization of enums by stringified member names.
2. It's pretty limited and can't be extended much. We could extend this in the future to support stuff like this
```solidity
/// @param number forge-json: hex
/// @param inner forge-json: flatten
struct SomeStruct {
    uint256 number;
    AnotherStruct inner; 
}
```

However, all of this can be added in the future by just allowing more formats for `typeDescription` param and adding helpers for generating new description formats.

**Comment by klkvr** [comment]
Added cheatcode for serialization and `forge bind-json` command. It can be used to generate `JsonBindings` library containing helpers for serialization/deserialization of all structs in the project.

Example output for project with 1 struct
```solidity
pragma solidity >=0.6.2 <0.9.0;
pragma experimental ABIEncoderV2;

import {Vm} from "forge-std/Vm.sol";
import {SomeStruct} from "src/SomeStruct.sol";

library JsonBindings {
    Vm constant vm = Vm(address(uint160(uint256(keccak256("hevm cheat code")))));

    string constant schema_SomeStruct = "SomeStruct(uint256 a,uint256 b)";

    function serialize(SomeStruct memory value) internal pure returns (string memory) {
        return vm.serializeJsonType(schema_SomeStruct, abi.encode(value));
    }

    function deserializeSomeStruct(string memory json) public pure returns (SomeStruct memory) {
        return abi.decode(vm.parseJsonType(json, schema_SomeStruct), (SomeStruct));
    }

    function deserializeSomeStruct(string memory json, string memory path) public pure returns (SomeStruct memory) {
        return abi.decode(vm.parseJsonType(json, path, schema_SomeStruct), (SomeStruct));
    }

    function deserializeSomeStructArray(string memory json, string memory path) public pure returns (SomeStruct[] memory) {
        return abi.decode(vm.parseJsonTypeArray(json, path, schema_SomeStruct), (SomeStruct[]));
    }
}

```

bindings are written to `utils/JsonBindings.sol` by default and can be later used like this:
```solidity
import "../utils/JsonBindings.sol";
import "forge-std/Test.sol";

struct SomeStruct {
    uint256 a;
    uint256 b;
}

contract SomeStructJsonTest is Test {
    using JsonBindings for *;

    function test() public {
        SomeStruct memory s = SomeStruct(1, 2);
        string memory json = s.serialize();
        SomeStruct memory s2 = JsonBindings.deserializeSomeStruct(json);
        assertEq(keccak256(abi.encode(s)), keccak256(abi.encode(s2)));
    }
}
```

The command itself is pretty straightforward - just walk the AST and write the methods, the only non-trivial part is preprocessing which need to obtain valid AST from solc even when bindings are outdated and would cause compiler to fail normally. We rely on solang parser for this.

**Review by mattsse — COMMENTED** [review]
nice, this is pretty cool!

still need to take a closer look but I believe this should work, and the helper commands should make this convenient to use.

if possible, we def want some forgetest! for this as well

File: crates/forge/bin/cmd/bind_json.rs [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+};
+
+foundry_config::impl_figment_convert!(BindJsonArgs, opts);
+
+/// CLI arguments for `forge bind-json`.
+#[derive(Clone, Debug, Parser)]
+pub struct BindJsonArgs {
+    #[command(flatten)]
+    opts: CoreBuildArgs,
+}
```
Comment (mattsse) [inline crates/forge/bin/cmd/bind_json.rs:32]: this this also include structs from dependencies?
I think we can make this optional

maybe even some glob filter, but this we can add in a followup

**Review by mattsse — CHANGES_REQUESTED** [review]
nice work!

only have a few questions/suggestions

File: crates/forge/bin/cmd/eip712.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    },
+    CompilerSettings,
+};
+use std::{collections::BTreeMap, path::PathBuf};
+
+foundry_config::impl_figment_convert!(Eip712Args, opts);
+
+/// CLI arguments for `forge eip712`.
+#[derive(Clone, Debug, Parser)]
+pub struct Eip712Args {
```
[the remark this answers is not in the document]
Comment (klkvr) [reply crates/forge/bin/cmd/eip712.rs:20]: we are mostly documenting subcommands at the level of `ForgeSubcommands` variants: https://github.com/foundry-rs/foundry/blob/e48776e7cee632130bea3c2165cf1d39fea39e98/crates/forge/bin/opts.rs#L168-L169

should I add docs to structs also?

**Comment by klkvr** [comment]
Added config section:
```
[bind_json]
# path for generated bindings file
out = "utils/JsonBindings.sol"
# array of globs to include, if empty defaults to all project sources, excluding library files
include = []
# array of globs to exclude 
exclude = []
```

File: crates/forge/bin/cmd/bind_json.rs [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+        for struct_to_write in &self.structs_to_write {
+            result.push_str(&format!(
+                "    string constant schema_{} = \"{}\";\n",
+                struct_to_write.name_in_fns, struct_to_write.schema
+            ));
+        }
+
+        // write serialization functions
+        for struct_to_write in &self.structs_to_write {
+            result.push_str(&format!(
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind_json.rs:501]: write!(

File: crates/forge/bin/cmd/bind_json.rs [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+}
+
+impl ResolvedState {
+    fn write(self) -> Result<String> {
+        let mut result = String::new();
+        self.write_imports(&mut result);
+        self.write_vm(&mut result);
+        self.write_library(&mut result);
+
+        fs::create_dir_all(self.target_path.parent().unwrap())?;
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind_json.rs:440]: if let Some(parent)

File: crates/forge/bin/cmd/bind_json.rs [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+
+        // Resolver for EIP712 schemas
+        let resolver = Resolver::new(&asts);
+
+        let mut structs_to_write = Vec::new();
+
+        let include =
+            config.bind_json.include.into_iter().map(GlobMatcher::new).collect::<Vec<_>>();
+        let exclude =
+            config.bind_json.exclude.into_iter().map(GlobMatcher::new).collect::<Vec<_>>();
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind_json.rs:307]: can we not put the struct in the config directly?

File: crates/forge/bin/cmd/eip712.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    #[command(flatten)]
+    opts: CoreBuildArgs,
+}
+
+impl Eip712Args {
+    pub fn run(self) -> Result<()> {
+        let config = self.try_load_config_emit_warnings()?;
+        let mut project = config.create_project(false, true)?;
+        let target_path = dunce::canonicalize(self.target_path)?;
+        project.settings.update_output_selection(|selection| {
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/eip712.rs:34]: If we're only reading one file can we just set the project to compile that?

Comment (klkvr) [reply crates/forge/bin/cmd/eip712.rs:34]: we're doing `.files([target_path.clone()])` below, or you mean something else?

File: crates/forge/bin/cmd/bind_json.rs (hunk 5 above)
Comment (klkvr) [reply crates/forge/bin/cmd/bind_json.rs]: yep, just needed manual `PartialEq` a227ed8

File: crates/forge/bin/cmd/eip712.rs (hunk 6 above)
Comment (DaniPopes) [reply crates/forge/bin/cmd/eip712.rs:34]: missed that

**Review by DaniPopes — APPROVED** [review]
lgtm pending @mattsse

**Review by mattsse — APPROVED** [review]
gg!

**Pull request merged**